pandas/core/groupby/generic.py

```diff
@@ -968,7 +968,8 @@ def aggregate(self, func=None, *args, **kwargs):
                     result = self._aggregate_frame(func)
                 else:
                     result.columns = Index(
-                        result.columns.levels[0], name=self._selected_obj.columns.name
+                        [i[:-1] if len(i) > 2 else i[0] for i in result.columns],
```

Comment:
what are you actaully trying to do here?

Reply:
@jreback Currently, when we get here, if we have
```
>>> result.columns
MultiIndex([('B', '<lambda>'),
            ('C', '<lambda>'),
            ('D', '<lambda>')],
           )
```
then the current line of code ignores the last level:
```
>>> Index(result.columns.levels[0], name=self._selected_obj.columns.name)
Index(['B', 'C', 'D'], dtype='object')
```

However, if (as in the new test I've added) we have
```
>>> result.columns
MultiIndex([(1, 3, '<lambda>'),
            (1, 4, '<lambda>'),
            (2, 3, '<lambda>'),
            (2, 4, '<lambda>')],
           )
```
then the current line of code will only take the first level:
```
>>> Index(result.columns.levels[0], name=self._selected_obj.columns.name)
Int64Index([1, 2], dtype='int64')
```

However, what we need is
```
MultiIndex([(1, 3),
            (1, 4),
            (2, 3),
            (2, 4)],
           )
```

So, that's what I'm trying to do here - if there's only two levels, return the first one as an Index (as is currently done on `master`), and if there's more levels, return all but the last one as a MultiIndex

Reply:
Maybe misreading but this entire expression seems strange. Why would master arbitrarily select the first column index level from the caller?

I think there's a more general fix to be had here

Reply:
> Why would master arbitrarily select the first column index level from the caller?

I think the idea is to select everything _except_ for the last level, which is the one containing the name of the function e.g. `'<lambda>'`

> I think there's a more general fix to be had here
Seems plausible :)

Reply:
@WillAyd @jreback thanks for your reviews - have found a cleaner way to write this:
```
                    result.columns = result.columns.rename(
                        [self._selected_obj.columns.name] * result.columns.nlevels
                    ).droplevel(-1)
```

Before I update the tests according the tests, could you let me know if this seems like the correct patch?

Reply:
Can you not just copy the columns and drop level thereafter?

Reply:
@WillAyd not sure I understand what you mean - as in,
```
result.columns.droplevel(-1).rename(self._selected_obj.columns.name)
```
?

The problem is that then, if we have
```
(Pdb) result.columns
MultiIndex([('B', '<lambda>'),
            ('C', '<lambda>'),
            ('D', '<lambda>')],
           )
```
then
```
(Pdb) result.columns.droplevel(-1)
Index(['B', 'C', 'D'], dtype='object')
```
_but_, if we have
```
(Pdb) result.columns
MultiIndex([(1, 3, '<lambda>'),
            (1, 4, '<lambda>'),
            (2, 3, '<lambda>'),
            (2, 4, '<lambda>')],
           )
```
then
```
(Pdb) result.columns.droplevel(-1)
MultiIndex([(1, 3),
            (1, 4),
            (2, 3),
            (2, 4)],
           )
```

In the first case, renaming can be done with e.g. a string:
```
result.columns.droplevel(-1).rename(self._selected_obj.columns.name)
```
but in the second case, it needs to be done with a list of strings:
```
result.columns.droplevel(-1).rename([self._selected_obj.columns.name]*2)
```